lib/filesystem.nix

```
@@ -372,46 +369,59 @@ in
     :::{.note}
     `a.nix` cannot directly take as inputs packages defined in a child directory, such as `b1`.
     :::
-
-    :::{.warning}
-    As of now, `lib.packagesFromDirectoryRecursive` cannot create nested scopes for sub-directories.
-
-    In particular, files under `b/` can only require (as inputs) other files under `my-packages`,
-    but not to those in the same directory, nor those in a parent directory; e.g, `b2.nix` cannot directly
-    require `b1`.
-    :::
     ::::
   */
   packagesFromDirectoryRecursive =
+    let
+      inherit (lib) concatMapAttrs id makeScope recurseIntoAttrs removeSuffix;
+      inherit (lib.path) append;
+
+      # Generate an attrset corresponding to a given directory.
+      # This function is outside `packagesFromDirectoryRecursive`'s lambda expression,
+      #  to prevent accidentally using its parameters.
+      processDir = { callPackage, directory, ... }@args:
+        concatMapAttrs (name: type:
+          # for each directory entry
+          let path = append directory name; in
+          if type == "directory" then {
+            # recurse into directories
+            "${name}" = packagesFromDirectoryRecursive (args // {
+              directory = path;
+            });
+          } else if type == "regular" && hasSuffix ".nix" name then {
+            # call .nix files
+            "${removeSuffix ".nix" name}" = callPackage path {};
+          } else if type == "regular" then {
+            # ignore non-nix files
+          } else throw ''
+            lib.filesystem.packagesFromDirectoryRecursive: Unsupported file type ${type} at path ${toString path}
+          ''
+        ) (builtins.readDir directory);
+    in
     {
       callPackage,
+      newScope ? throw "lib.packagesFromDirectoryRecursive: newScope wasn't passed in args",
```

Can we default to the newScope from callPackage?

```suggestion
      newScope ? (callPackage ({ newScope }: { inherit newScope; }) { }).newScope,
```

---

@katexochen Do you want to take that over. It seems the PR maintainer is unresponsive. Can you squash the original commits into a new pr and link this one so we can move on?

---

Regarding your suggestion. I'd prefer to merge what is already reviewed. Its easy to make a follow up pr for that.

---

Let's give them a day or two to react. Otherwise I'll take over.

---

Yes, I'll take over, but it's a bit fiddly, the fixups don't squash cleanly and I have to resolve merge conflicts with master, but I'm on it.

---

@nbraud looking at the code now, I think my suggestion doesn't make any sense as this throw is never triggered? The way I understand it, if now newScope is passed we will not use nested scopes, so there is no point of providing a default for newScope.